ctk/common/src/main/kotlin/org/codice/compliance/verification/profile/SingleSignOnProfileVerifier.kt

```diff
@@ -209,4 +213,67 @@ class SingleSignOnProfileVerifier(private val authnRequest: AuthnRequest,
         }
         return audienceRestriction.first()
     }
+
+    /**
+     * 3.1 Holder of Key
+     */
+    internal fun verifyHolderOfKey() {
+        val holderOfKeyList = samlResponseDom.recursiveChildren(SUBJECT_CONFIRMATION)
+            .filter { it.attributeText("Method") == HOLDER_OF_KEY_URI }
+
+        if (holderOfKeyList.isEmpty()) return
+
+        val holderOfKeyDataList = holderOfKeyList
+            .flatMap { it.children(SUBJECT_CONFIRMATION_DATA) }
+
+        if (holderOfKeyDataList.isEmpty())
+            throw SAMLComplianceException.create(SAMLProfiles_3_1_a,
```

Comment:
❗️ Technically this would pass if we had multiple `Holder-of-Key`'s and only one of them had a `<SubjectConfirmationData>`. The quick fix is `holderOfKeyDataList.size() == holderOfKeyList.size()`, but I personally think this would read better:
```
private fun verifyHolderOfKey() {
    val holderOfKeyList = node.recursiveChildren(SUBJECT_CONFIRMATION)
            .filter { it.attributeText("Method") == HOLDER_OF_KEY_URI }

    if (holderOfKeyList.isEmpty()) return

    holderOfKeyList.forEach {
        val subjectConfirmationDataElements = it.children("SubjectConfirmationData")
        if (subjectConfirmationDataElements.isEmpty())
            throw SAMLComplianceException.create(SAMLProfiles_3_1_a,
                    message = "<SubjectConfirmationData> not found within Holder of Key " +
                            "<SubjectConfirmation>",
                    node = it)

        subjectConfirmationDataElements.forEach {
            verifyHolderOfKeyData(it)
        }
    }
}
```

Reply:
Just had the thought that maybe we should add a unit test for two HolderOfKeys in the Response. That would've caught this bug.